File: matlab/tools/connectivity/pearson.m
function avg_pc = pearson(values,fs,varargin)
% Calculate the Pearson correlation coefficients between channels in iEEG data.
%
% Parameters:
% - values (matrix): Matrix representing iEEG data. Each column is a channel, and each row is a time point.
% - fs (float): Sampling frequency of the iEEG data.
% - do_tw (logical, optional): If true, calculate windowed correlations; if false, calculate overall correlations.
% - tw (float, optional): Size of the time window in seconds for windowed correlation calculation.
%
% Returns:
% - avg_pc (matrix): Pearson correlation coefficients between channels. If windowed, returns an average over time windows.
%
% Example:
% values = randn(100, 5);  % Replace with your actual data
% fs = 250;
% correlations = pearson(values, fs);
% correlations = pearson(values, fs, true);
% correlations = pearson(values, fs, true, 3);

defaults = {false,2};

for i = 1:length(varargin)
    if isempty(varargin{i})
        varargin{i} = defaults{i};
    end
end

p = inputParser;
addRequired(p, 'values', @isnumeric);
addRequired(p, 'fs', @isnumeric);
addOptional(p, 'do_tw', defaults{1}, @(x) ismember(x,[0,1]));
addOptional(p, 'tw', defaults{2}, @isnumeric);

parse(p, values, fs, varargin{:});

% Access parsed values
values = p.Results.values;
fs = p.Results.fs;
tw = p.Results.tw;
do_tw = p.Results.do_tw;

nchs = size(values,2);

if ~do_tw
    avg_pc = corrcoef(values);
else
    % Define time windows
    iw = round(tw*fs);
    window_start = 1:iw:size(values,1);

    % remove dangling window
    if window_start(end) + iw > size(values,1)
        window_start(end) = [];
    end
    nw = length(window_start);

    % initialize output vector
    all_pc = nan(nchs,nchs,nw);

    % Calculate pc for each window
    for i = 1:nw

        % Define the time clip
        clip = values(window_start:window_start+iw,:);


        pc = corrcoef(clip);
        %pc(logical(eye(size(pc)))) = 0;

        % unwrap the pc matrix into a one dimensional vector for storage
        all_pc(:,:,i) = pc;

    end

    % Average the network over all time windows
    avg_pc = mean(all_pc,3,'omitnan');

end
end

File: python/CNTtools/tools/pearson.py
import numpy as np
from beartype import beartype
from numbers import Number


@beartype
def pearson(values: np.ndarray, fs: Number, win: bool, win_size: Number) -> np.ndarray:
    """
    Calculate the Pearson correlation coefficients between channels in iEEG data.

    Parameters:
    - values (np.ndarray): 2D array representing iEEG data. Each column is a channel, and each row is a time point.
    - fs (float): Sampling frequency of the EEG data.
    - win (bool): If True, calculate windowed correlations; if False, calculate overall correlations.
    - win_size (float): Size of the time window in seconds for windowed correlation calculation.

    Returns:
    - np.ndarray: Pearson correlation coefficients between channels. If windowed, returns an average over time windows.

    Examples:
    >>> values = np.random.rand(100, 5)
    >>> fs = 250
    >>> win_size = 2
    >>> win = True
    >>> correlations = pearson(values, fs, win, win_size)
    """

    nchs = values.shape[1]

    if not win:
        avg_pc = np.corrcoef(values, rowvar=False)
    else:
        # Define time windows
        iw = round(win_size * fs)
        window_start = np.arange(0, values.shape[0], iw)

        # Remove dangling window
        if window_start[-1] + iw > values.shape[0]:
            window_start = window_start[:-1]

        nw = len(window_start)

        # Initialize output array
        all_pc = np.empty((nchs, nchs, nw))
        all_pc[:] = np.nan

        # Calculate pc for each window
        for i, start in enumerate(window_start):
            # Define the time clip
            clip = values[start : start + iw, :]

            pc = np.corrcoef(clip, rowvar=False)
            # np.fill_diagonal(pc, 0)

            # Unwrap the pc matrix into a one-dimensional vector for storage
            all_pc[:, :, i] = pc

        # Average the network over all time windows
        avg_pc = np.nanmean(all_pc, axis=2)

    return avg_pc
